src/main/scala/org/kframework/kore/interfaces/Interfaces.scala

```diff
@@ -481,61 +481,251 @@ object pattern {
     def unapply(arg: Equals): Option[(Pattern, Pattern)] = Some(arg._1, arg._2)
   }
 
+
 }
 
-/**
-  * Provides a Builder type that allows building Pattern types in [[pattern]].
-  *
-  * Sample Usage -
-  *
-  * Given a concrete implementation of [[build.Builders]], one can create new patters -
-  * {{{
-  *   /* Given Concrete Implementation */
-  *   val builder: Builders = ConcreteBuilders
-  *   /* A Pattern Or(X:Int, Y:Int) can be constructed in the following way */
-  *   val or: Or = builder.Or(builder.Variable("X", Sort("Int")), builder."Y", Sort("Int"))
-  * }}}
-  *
-  */
-object build {
+object outer {
 
-  import pattern._
+  import org.kframework.kore.interfaces.{pattern => p}
+
+  type Attributes = Seq[p.Pattern]
+
+  /**
+    * Trait That is Extended by Every Outer Construct.
+    */
+  trait Outer {
+    val att: Attributes
+    val sorts: Set[p.Sort]
+    val symbols: Set[p.Symbol]
+    val sentences: Seq[Sentence]
+
+    def onAttributes(f: p.Pattern => p.Pattern): Outer
+
+    def getBySymbol(key: p.Symbol): Seq[Seq[p.Pattern]] = att.collect({ case p.Application(`key`, args) => args })
+  }
+
+
+  /** Kore Definition.
+    *
+    * Provides (Implementation for members)
+    *   - sorts, a set of all [[pattern.Sort]]s, obtained recursively.
+    *   - symbols, a set of all [[pattern.Symbol]]s, obtained recursively.
+    *   - sentences, a set of all [[Sentence]]s, obtained recursively.
+    *
+    *   Requires (Implementation for members)
+    *   - modules, a Seq of [[Module]]s
+    */
+  trait Definition extends Outer {
+    val modules: Seq[Module]
+
+    // Derived operations
+    lazy val sorts: Set[p.Sort] = modules.flatMap(_.sorts).toSet
+    lazy val symbols: Set[p.Symbol] = modules.flatMap(_.symbols).toSet
+    lazy val sentences: Seq[Sentence] = modules.flatMap(_.sentences)
+
+    override def onAttributes(f: p.Pattern => p.Pattern): Definition
+  }
+
+  object Definition {
+    def unapply(arg: Definition): Option[(Seq[Module], Attributes)] = Some(arg.modules, arg.att)
+  }
+
+  /** Kore Module.
+    *
+    * Provides (Implementation for members)
+    *   - sorts, a set of all [[pattern.Sort]]s, obtained recursively.
+    *   - symbols, a set of all [[pattern.Symbol]]s, obtained recursively.
+    *
+    *   Requires (Implementation for members)
+    *   - modules, a Seq of [[Sentence]]s
+    */
+  trait Module extends Outer {
+    val name: p.Name
+    val sentences: Seq[Sentence]
+
+    // Derived operations
+    lazy val sorts: Set[p.Sort] = sentences.flatMap(_.sorts).toSet
+    lazy val symbols: Set[p.Symbol] = sentences.flatMap(_.symbols).toSet
```

Comment:
so `symbols` collects *both* the declared and the used symbols?  For example, if one mistypes one symbol in a sentence, then that miss-typed one is also collected?  What't the use of such a method, if any?  The more I look over this code the more I feel that we should do the separation between Main and Extended KORE interfaces.  The Main one should be a mirror of the KORE language syntax that users who what to use K programmatically can use.